pkg/agent/api/delegatedidentity/v1/service.go

```diff
@@ -0,0 +1,228 @@
+func RegisterService(s *grpc.Server, service *Service) {
+	delegatedidentityv1.RegisterDelegatedIdentityServer(s, service)
+}
+
+type Attestor interface {
+	Attest(ctx context.Context) ([]*common.Selector, error)
+}
+
+type Config struct {
+	Log             logrus.FieldLogger
+	Manager         manager.Manager
+	Attestor        workload_attestor.Attestor
+	AuthorizedUsers []string
+}
+
+func New(config Config) *Service {
+	authorizedUsers := map[string]bool{}
+
+	for _, user := range config.AuthorizedUsers {
+		authorizedUsers[user] = true
+	}
+
+	return &Service{
+		manager:         config.Manager,
+		attestor:        endpoints.PeerTrackerAttestor{Attestor: config.Attestor},
+		authorizedUsers: authorizedUsers,
+	}
+}
+
+// Service implements the privileged server
+type Service struct {
+	delegatedidentityv1.UnsafeDelegatedIdentityServer
+
+	manager  manager.Manager
+	attestor Attestor
+
+	// Spiffe IDs of users that are authorized to use this API
+	authorizedUsers map[string]bool
+}
+
+// isCallerAuthorized attests the caller and returns true if its identity is on
+// the authorized users map.
+func (s *Service) isCallerAuthorized(ctx context.Context) (bool, error) {
+	callerSelectors, err := s.attestor.Attest(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	identities := s.manager.MatchingIdentities(callerSelectors)
+
+	for _, identity := range identities {
+		id := identity.Entry.SpiffeId
+
+		if _, ok := s.authorizedUsers[id]; ok {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (s *Service) SubscribeToX509SVIDs(req *delegatedidentityv1.SubscribeToX509SVIDsRequest, stream delegatedidentityv1.DelegatedIdentity_SubscribeToX509SVIDsServer) error {
+	ctx := stream.Context()
+	log := rpccontext.Logger(ctx)
+
+	authorized, err := s.isCallerAuthorized(ctx)
+
+	if err != nil {
+		log.WithError(err).Error("Workload attestation failed")
+		return status.Error(codes.Internal, "workload attestation failed")
+	}
+
+	if !authorized {
+		return status.Error(codes.PermissionDenied, "no authorized")
+	}
+
+	selectors, err := api.SelectorsFromProto(req.Selectors)
+	if err != nil {
+		return err
+	}
+
+	subscriber := s.manager.SubscribeToCacheChanges(selectors)
+	defer subscriber.Finish()
+
+	for {
+		select {
+		case update := <-subscriber.Updates():
+			if err := sendX509SVIDResponse(update, stream, log); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func sendX509SVIDResponse(update *cache.WorkloadUpdate, stream delegatedidentityv1.DelegatedIdentity_SubscribeToX509SVIDsServer, log logrus.FieldLogger) (err error) {
+	resp, err := composeX509SVIDBySelectors(update)
+	if err != nil {
+		log.WithError(err).Error("Could not serialize X.509 SVID response")
+		return status.Error(codes.Unavailable, "could not serialize response")
+	}
+
+	if err := stream.Send(resp); err != nil {
+		log.WithError(err).Error("Failed to send X.509 SVID response")
+		return err
```

Comment:
I'm not entirely sure under what circumstances we'd end up here, or who might actually receive this return value (and thus if it should be formatted as a grpc status code or not).

@azdagron might know the right thing to do. For the sake of consistency, a code feels like the right move, I'm not sure if codes.Internal or codes.Unavailable is more appropriate

Reply:
We haven't typically converted stream send errors into gRPC statuses since they obviously aren't ever seen by the other end and usually render pretty unambiguously in logs and what not.